src/esp/scene/SceneNode.cpp

```diff
@@ -16,16 +18,61 @@ SceneNode::SceneNode(SceneNode& parent)
   setId(parent.getId());
   setCachedTransformations(Mn::SceneGraph::CachedTransformation::Absolute);
   absoluteTransformation_ = absoluteTransformation();
+  nodeSensorSuite_ = new esp::sensor::SensorSuite(*this);
+  subtreeSensorSuite_ = new esp::sensor::SensorSuite(*this);
 }
 
 SceneNode::SceneNode(MagnumScene& parentNode)
     : Mn::SceneGraph::AbstractFeature3D{*this} {
   setParent(&parentNode);
   setCachedTransformations(Mn::SceneGraph::CachedTransformation::Absolute);
   absoluteTransformation_ = absoluteTransformation();
+  nodeSensorSuite_ = new esp::sensor::SensorSuite(*this);
+  subtreeSensorSuite_ = new esp::sensor::SensorSuite(*this);
+}
+
+SceneNode::~SceneNode() {
+  // if the entire scene graph is being deleted, no need to update anything
+  if (SceneGraph::isRootNode(*this)) {
+    return;
+  }
+  // if it is to delete a subtree
+  // then two cases:
+  // 1) current node is the root node (of this subtree).
+  // In case 1, you do not have to do anything since its parent node is nullptr
+  // already!! This is because magnum destroys the nodes in a top-down manner
+  // recursively, root node first, and then its child nodes
+  if (!this->parent()) {
+    return;
+  }
+
+  // 2) current node is NOT the root node.
+  // In case 2, you need update the sensorSuites stored in each ancestor node in
+  // a bottom-up manner.
+
+  // First erase all nodes in nodeSensorSuite from its parent's nodeSensorSuite
+  for (const auto& sensor : nodeSensorSuite_->getSensors()) {
+    static_cast<SceneNode*>(this->parent())
+        ->getNodeSensorSuite()
```

Comment:
Should this be the subtreeSensorSuite?

Reply:
@erikwijmans : in the `nodeSensorSuite`, it stores the sensors owned by this node. For example, if this node represents "left arm" and owns 3 sensors (in 3 child, LEAF nodes), then these 3 sensors will be in this `nodeSensorSuite` and it does not include sensors in the subtrees, e.g., sensors on the "fingers". 

Now suppose you are to delete 1 of the 3 sensors. We do so by deleting that sensor node (the sensor "holder" node). In this case, this for loop is to help get rid of the sensor in its actual "owner".

Reply:
I think I get how this works without being a nightmare after looking more in depth.  You are stopping a sensor from being attached to a sensor by tagging that scene node as a permanent leaf.  

I am trying to think of how to make this more clear in the code.  Maybe adding some checks that must be true due to this logic?  If I am understanding this correctly, this loop is unnecessary for non-leaves as a node's parent will only have its child sensor in its sensor suite if that child is a leaf?  A corollary is that non-leaves do not own sensors do not own sensors, so the sensorSuite for a non-leaf should be empty upon destruction?

Reply:
Yes, that's correct. 
- Sensors can only be attached to leaf nodes.
- For non-leaves, the loop is unnecessary because it will own no sensors that could be in its parent's nodeSensorSuite.
- Non-leaves do not own sensors, but sensors could be attached to its direct child nodes, in which case they will be included in its nodeSensorSuite. (Otherwise, nodeSensorSuites would only ever contain 0 or 1 sensors)